### PR TITLE
Auto nesting of input arguments.

### PR DESCRIPTION
**Pull request by yfried-redhat**
RHOSINFRA-63 RHOSINFRA-22 Auto nesting of input arguments.

Arguments will now populate the settings tree automatically by mapping
of argument name to a nested dict.

Example:
The input "$ ir-COMMAND SUBCOMMAND --arg-name=arg-value" will be map:

```yaml
    COMMAND:
        arg:
            name: "arg-value"
```
That nested dict will be merged into the settings YAML found in:

  settings/COMMAND/SUBCOMMAND/SUBCOMMAND.yaml

Based on the spec file:

  settings/COMMAND/SUBCOMMAND/SUBCOMMAND.spec

Adds ValueArgument to clg types. Arguments of these type can reslove
defaults according to documentation: If not provided in CLI, search in
ENV vars and later in given ini file.

Adds YamlFileArgument that will load YAMLS from given paths directly as
dict. Default search path for the argument from the above example would
be:

  $ ir-COMMAND SUBCOMMAND --arg-name=path/to/file.yaml
  settings_dir/COMMAND/SUBCOMMAND/arg/name/path/to/file/yaml

Adds search_tree method to search all values of a given key inside a
nested dict tree.
Adds unittest for method

### CODE REVIEW

**Comment by yfried-redhat** [comment]
@obaranov @aopincar @tkammer please review

**Comment by yfried-redhat** [comment]
tested with: https://rhos-jenkins.rhev-ci-vms.eng.rdu2.redhat.com/view/RHOS/view/RHOS7/job/poc_yfried_dynamic_topology/

File: cli/provision.py [hunk 1]
```diff
@@ -44,13 +50,13 @@ def process_topology_args(clg_args, app_settings_dir):
...
-                topology_dict[node_type] = yaml.load(stream)
-                topology_dict[node_type]['amount'] = int(number)
+            topology_dict[node_type] = spec.find_file(node_type + ".yaml",
+                                                      topology_dir)
+            topology_dict[node_type]['amount'] = int(number)
+
+        self.value = topology_dict
 
-        clg_args['topology'] = topology_dict
+clg.TYPES.update({'Topology': TopologyArgument})
```
Comment (obaranov) [inline cli/provision.py]: I'd suggest to have  types update somewhere in one place. Now we have update in provision.py and in the spec.py file.

File: settings/provisioner/virsh/virsh.spec [hunk 2]
```diff
@@ -3,38 +3,38 @@ subparsers:
     virsh:
         help: Provision systems using virsh
         groups:
-            - title: host
+            - title: Hypervisor
               options:
-                  host:
-                      type: str
+                  host-address:
+                      type: Value
```
Comment (obaranov) [inline settings/provisioner/virsh/virsh.spec]: Unable to override default values for the custopm clg type in the ini file provided with --from-file argument. For example: 
I have in ini file
topology-network=default.yml
topology-nodes=1_controller
host-address=rhos-qe-06.lab.eng.brq.redhat.com

But provisioner is still using default values. 

The same is mostly likely actual for environment variables.

File: cli/provision.py (hunk 1 above)
Comment (yfried-redhat) [inline cli/provision.py]: I agree, but the question is:
Should TopologyArgument be defined in spec or in provision?

File: settings/provisioner/virsh/virsh.spec (hunk 2 above)
Comment (yfried-redhat) [inline settings/provisioner/virsh/virsh.spec]: Will look into that.
We need unittest for the gate.

File: cli/provision.py [hunk 3]
```diff
@@ -238,9 +183,14 @@ def run(self):
 
     def collect_settings(self):
         settings_files = self.sub_command.get_settings_files()
-        settings_dict = self.sub_command.get_settings_dict()
+        arguments_dict = {APPLICATION: spec.get_arguments_dict(self.args)}
```
Comment (obaranov) [inline cli/provision.py]: I think the get_arguments_dict should be moved out of spec file. Application should transform spec to the dict. Spec.py should don't care about required format for application.

File: cli/spec.py [hunk 4]
```diff
@@ -16,30 +20,187 @@
...
+    settings_dir = None
+    subcommand = None
+
+    def __init__(self, value=None, required=None):
+        self.value = value
+        self.arg_name = None
+        self.required = None
+
+    @classmethod
+    def get_app_attr(cls, attribute):
```
Comment (obaranov) [inline cli/spec.py]: These classmethod's  in the ValueArgument are over-responsibility for that class which is the additional  clg type. I'd suggest to move these helper methods to something like Arguments Helper. This is up to you.

File: cli/provision.py (hunk 1 above)
Comment (obaranov) [inline cli/provision.py]: I vote to move TopologyArgument into spec.py

File: cli/spec.py [hunk 5]
```diff
@@ -16,30 +20,187 @@
...
+
+    def resolve_value(self, arg_name, defaults=None):
+        super(YamlFileArgument, self).resolve_value(arg_name, defaults)
+        search_first = os.path.join(self.get_app_attr("settings_dir"),
+                                    self.get_app_attr("subcommand"),
+                                    *arg_name.split("-"))
+        self.value = find_file(self.value, search_first)
+
+
+def find_file(filename, search_first):
```
Comment (obaranov) [inline cli/spec.py]: Maybe move this inside of YamlFileArgument and exetend TopologyArgument from YamlFileArgument. Just an idea.

File: cli/provision.py [hunk 6]
```diff
@@ -14,29 +13,36 @@
...
-    :param clg_args: Dictionary based on cmd-line args parsed by clg
-    :param app_settings_dir: path to the base directory holding the
-        application's settings. App can be provisioner\installer\tester
-        and the path would be: settings/<app_name>/
-    """
+    def resolve_value(self, arg_name, defaults=None):
+        """
+        Merges topology files in a single topology dict.
+
+        :param clg_args: Dictionary based on cmd-line args parsed by clg
```
Comment (obaranov) [inline cli/provision.py]: Param names should be updated in the documentation.

File: tests/test_provision.py [hunk 7]
```diff
@@ -28,11 +28,13 @@ def test_dynamic_topology(tmpdir):
 """)
     # prepare config
     app_path = os.path.join(root_dir.strpath, "..")
-    res_args = dict(topology="10_controller,2_compute")
 
+    provision.TopologyArgument.settings_dir = app_path
```
Comment (obaranov) [inline tests/test_provision.py]: Setting the "settings_dir" for the TopologyArgument is required but hidden from user. Can we move settings_dir as an optional argument to the constructor?

**Comment by obaranov** [comment]
Review completed from my side.

File: cli/spec.py (hunk 4 above)
Comment (yfried-redhat) [inline cli/spec.py]: I'm not sure how to do that.
This method is setting and getting class attributes, but it's called from both classmethods and instance methods, so the only way I know to do that is with a classmethod

File: cli/spec.py (hunk 5 above)
Comment (yfried-redhat) [inline cli/spec.py]: I think I'll move that to utils. It's not tightly bound to the class

File: tests/test_provision.py (hunk 7 above)
Comment (yfried-redhat) [inline tests/test_provision.py]: ok

File: cli/provision.py (hunk 6 above)
Comment (yfried-redhat) [inline cli/provision.py]: done

File: cli/provision.py (hunk 3 above)
Comment (yfried-redhat) [inline cli/provision.py]: done

File: settings/provisioner/virsh/virsh.spec (hunk 2 above)
Comment (yfried-redhat) [inline settings/provisioner/virsh/virsh.spec]: solved

File: cli/provision.py (hunk 1 above)
Comment (yfried-redhat) [inline cli/provision.py]: done

**Comment by yfried-redhat** [comment]
@obaranov pushed fixed. please review. tnx

**Comment by obaranov** [comment]
@yfried-redhat looks good for me

**Pull request merged**